src/QuadVelocityController.cpp

```diff
@@ -215,150 +250,103 @@ bool QuadVelocityController::getTransformAtTime(
-    {
-        ros::Time last_time = time
-                            - ros::Duration(1.0 / expected_update_frequency_);
-
-        // Get a brand new transform
-        // Blocks until able to transform at the requested time
-        bool fetched_transform = getTransformAtTime(last_transform_stamped_,
-                                                    last_time,
-                                                    timeout);
-
-        // Check if the call failed
-        if (!fetched_transform)
-        {
-            ROS_ERROR("Failed to fetch first transform in QuadVelocityController");
-            have_last_transform_ = false;
-            return false;
-        }
-
-        have_last_transform_ = true;
-    }
-
-    // Get a brand new transform
-    // Blocks until able to transform at the requested time
-    geometry_msgs::TransformStamped transform_stamped;
-    bool fetched_transform = getTransformAtTime(transform_stamped,
-                                                time,
-                                                timeout);
-
-    // Check if the call failed
-    if (!fetched_transform)
+    // Wait until there's a message with stamp >= time
+    if (!waitForOdometryAtTime(time, timeout))
     {
-        ROS_ERROR("Failed to fetch new transform in QuadVelocityController");
-        have_last_transform_ = false;
+        ROS_ERROR("Timed out waiting for odometry message in getVelocityAtTime");
         return false;
     }
 
-    // Make sure the difference between transforms isn't too much
-    // Calculate the lastest time between transforms that is allowed.
-    ros::Time latest_time_allowed = last_transform_stamped_.header.stamp
-                + ros::Duration(MAX_TRANSFORM_DIFFERENCE_SECONDS);
-    // Make the sure the transform didn't come after the last time
-    // that was acceptable.
-    if (transform_stamped.header.stamp > latest_time_allowed)
-    {
-        ROS_ERROR_STREAM(
-            "First available transform came after latest allowed time ("
-         << "transform time: " << transform_stamped.header.stamp << ", "
-         << "latest allowed time: " << latest_time_allowed
-         << ")");
-        return false;
-    }
-
-    return_velocities = twistFromTransforms(last_transform_stamped_,
-                                            transform_stamped);
-
-    // Store the old transform
-    last_transform_stamped_ = transform_stamped;
-
+    // Get first msg that is >= the requested time
+    std::vector<nav_msgs::Odometry>::iterator geq_msg
+        = std::upper_bound(odometry_msg_queue_.begin(),
+                           odometry_msg_queue_.end(),
+                           time,
+                           &QuadVelocityController::timeVsMsgStampedComparator
+                                <nav_msgs::Odometry>);
+
+    // Linear interpolation between next_odom and last_odom
+    const nav_msgs::Odometry& next_odom = *geq_msg;
+    const nav_msgs::Odometry& last_odom = *(geq_msg - 1);
```

Comment:
I think it would be good to check geq_msg isn't std:begin before doing this. Even if the logic of the class is supposed to guarantee that this doesn't turn into std::begin what if its accidentally broken in the future?